### DISCUSSION TITLE
string-ts roadmap

### DISCUSSION THREAD

**Discussion by gustavoguichard** [gustavoguichard/string-ts#36] [Ideas]
# Roadmap for v1

As I was asked by @p9f I don't have a specific Roadmap for string-ts yet, but will try to outline my ideas and people can contribute with ideas to this discussion so I can update the list.

Based on [MDN's list of String.prototype functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String) I got some inspiration for the useful native counterparts. I don't think we should implement every function just for the sake of it, though.
Here follows a list where:
✅ = implemented
◻️ = not yet implemented
🟠 = partially implemented
❓ = not sure if it makes sense
🚫 = not gonna implement

---

- ❓ [String.prototype.at()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/at) - unsure as charAt is similar and has better support, we could just implement negative indexes to it
- ✅ [String.prototype.charAt()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/charAt)
- 🚫 [String.prototype.charCodeAt()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/charCodeAt)
- 🚫 [String.prototype.codePointAt()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/codePointAt)
- ✅ [String.prototype.concat()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/concat)
- ✅ [String.prototype.endsWith()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/endsWith)
- 🚫 [String.fromCharCode()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/fromCharCode)
- 🚫 [String.fromCodePoint()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/fromCodePoint)
 - ✅ [String.prototype.includes()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes)
- ❓ [String.prototype.indexOf()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/indexOf)
- 🚫 [String.prototype.isWellFormed()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/isWellFormed)
- ❓ [String.prototype.lastIndexOf()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/lastIndexOf)
- 🚫 [String.prototype.localeCompare()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare)
- 🚫 [String.prototype.match()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match)
- 🚫 [String.prototype.matchAll()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/matchAll)
- 🚫 [String.prototype.normalize()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize)
- ✅ [String.prototype.padEnd()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padEnd)
- ✅ [String.prototype.padStart()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padStart)
- ✅ [String.prototype.repeat()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/repeat)
- 🟠 [String.prototype.replace()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace) - using Regex will result into the loose `string` type
- 🟠 [String.prototype.replaceAll()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replaceAll) - using Regex will result into the loose `string` type
- 🚫 [String.prototype.search()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/search)
- ✅ [String.prototype.slice()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/slice)
- ✅ [String.prototype.split()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/split)
- ✅ [String.prototype.startsWith()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith)
- 🚫 [String.prototype.substring()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substring) - `slice` may be enough
- 🚫 [String.prototype.toLocaleLowerCase()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toLocaleLowerCase)
- 🚫 [String.prototype.toLocaleUpperCase()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toLocaleUpperCase)
- ✅ [String.prototype.toLowerCase()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toLowerCase)
- 🚫 [String.prototype.toString()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toString)
- ✅ [String.prototype.toUpperCase()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toUpperCase)
- 🚫 [String.prototype.toWellFormed()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toWellFormed)
- ✅ [String.prototype.trim()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trim)
- ✅ [String.prototype.trimEnd()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trimEnd)
- ✅ [String.prototype.trimStart()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trimStart)
- 🚫 [String.prototype.valueOf()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/valueOf)

Other common string methods:
- ✅ camelCase
- ✅ constantCase
- ✅ delimiterCase
- ✅ kebabCase
- ✅ lowerCase
- ✅ pascalCase
- ✅ snakeCase
- ✅ titleCase
- ✅ upperCase
- ✅ [truncate](https://lodash.com/docs/4.17.15#truncate)
- ❓ [dedent](#14)

Remember that the aim of this library is to have the same expected result at both type and runtime levels, so if you are going to implement one of these methods and you can only do it partially, make sure you modify the runtime implementation to behave the same as the type-level. Here is an (outdated because I solved it) [example of it](https://github.com/gustavoguichard/string-ts/blob/2b6357d37aceee28c112bff307d672d1badc3744/src/primitives.ts#L195-L196).

**Comment by gustavoguichard** [comment]
# Roadmap for v2

Here are the decisions we took too early that will cause breaking changes so they'll have to be a major bump.

## Differentiate between casing functions that rely on the `words` method from those that are native
From the issue #35 

Like in [lodash](https://lodash.com/docs/4.17.15#lowerCase), we should differentiate the methods like `camelCase` - that should probably not be prefixed with `to` (e.g: `toCamelCase`) that will omit separators - from methods like `toLowercase` - which replicate the native behavior that will not omit separators.

This will require us to implement new methods `lowerCase` and `upperCase`, as well as renaming the current ones to `camelCase`, `constantCase`, etc. We'll just keep the `toLowercase` and `toUppercase` methods as they are.

We should also clarify that on the docs.


## Tree-shaking
To reduce the impact of this library in the bundle size we should strive for enabling tree-shaking or we should even consider doing isolated exports like lodash does:
```ts
import pick from 'lodash/pick'
```

As we started supporting more methods the size of this library surpassed 1kb gzipped and that should be reduced.

**Reply by p9f** [reply]
about the renaming - as far as I understand it would not need to change the behaviour of any existing function name, right?

So maybe we could do it in a minor version - deprecating (but not removing) the old functions?

**Reply by gustavoguichard** [reply]
Good idea @p9f !
We can start deprecating right away, both types and runtime functions.
